course3/homework-unique_ptr/test.cc

```diff
@@ -0,0 +1,136 @@
+#include "unique_ptr.h"
+
+#include <cassert>
+#include <iostream>
+#include <memory>
+
+
+struct Number
+{
+	int n;
+	static int ctr;
+	Number(int n) : n(n) { ctr++; }
+	Number(int n1, int n2) : n(n1 + n2) { ctr++; }
+	Number(const Number& n) : n(n.n) { ctr++; }
+	Number(Number& n) : n(n.n * 10) { ctr++; }
+	~Number() { ctr--; }
+};
+int Number::ctr = 0;
+
+void unique_ptr_test()
```

Comment:
Pick two types you want to point to, define two types pointers. You can use a type to group it.
```
struct std_unique_ptr_test
{
   using int_ptr = std::unique_ptr<int>;
   using number_ptr = std::unique_ptr<Number>;
   template<typename T, typename.. Args>
   std::unique_ptr<T> MakeUnique(Args&& ...args) { return std::make_unique<T>(std::forward(args)...)); }
};
struct karun_unique_ptr_test
{
   using int_ptr = karun::unique_ptr<int>;
   using number_ptr = karun::unique_ptr<Number>;
   template<typename T, typename.. Args>
   static karun::unique_ptr<T> MakeUnique(Args&& ...args) { return karun::make_unique<T>(std::forward(args)...)); }
};

template<typename Test>
void test_unique_ptr()
{
   Test::number_ptr u1;
   auto u2 = Test::make_unique<Number>(u1->n * 10);
   // ...
}

// ...
test_unique_ptr<std_unique_ptr_test>();
test_unique_ptr<karun_unique_ptr_test>();